From 75eb7055fa4cd61334b9e83ce39e373cc2b1047a Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Fri, 7 Mar 2025 14:42:25 +0100
Subject: [PATCH] topic: Boot from QSPI using UBI root filesystem

Configure the topic-miami board to be able to boot from QSPI NOR flash.
Synchronize the topic_miami_defconfig.

Signed-off-by: Mike Looijmans <mike.looijmans@topic.nl>
Upstream-Status: Pending
---
 arch/arm/dts/zynq-topic-miami.dts | 33 ++++++++++++++++---------------
 configs/topic_miami_defconfig     | 16 +++++++++++++++
 include/configs/topic_miami.h     | 25 ++++++++++++++---------
 3 files changed, 49 insertions(+), 25 deletions(-)

diff --git a/arch/arm/dts/zynq-topic-miami.dts b/arch/arm/dts/zynq-topic-miami.dts
index 8307a2ef9d..29f0a4f828 100644
--- a/arch/arm/dts/zynq-topic-miami.dts
+++ b/arch/arm/dts/zynq-topic-miami.dts
@@ -11,6 +11,10 @@
 	model = "Topic Miami Zynq Board";
 	compatible = "topic,miami", "xlnx,zynq-7000";
 
+	config {
+		u-boot,spl-payload-offset = <0x20000>;
+	};
+
 	aliases {
 		serial0 = &uart0;
 		spi0 = &qspi;
@@ -35,6 +39,7 @@
 	status = "okay";
 	num-cs = <1>;
 	flash@0 {
+		bootph-all;
 		compatible = "st,m25p80", "n25q256a", "jedec,spi-nor";
 		m25p,fast-read;
 		reg = <0x0>;
@@ -44,24 +49,12 @@
 		#address-cells = <1>;
 		#size-cells = <1>;
 		partition@0 {
-			label = "qspi-u-boot-spl";
-			reg = <0x00000 0x10000>;
-		};
-		partition@10000 {
-			label = "qspi-u-boot-img";
-			reg = <0x10000 0x60000>;
+			label = "qspi-boot-bin";
+			reg = <0x00000 0x100000>;
 		};
-		partition@70000 {
-			label = "qspi-device-tree";
-			reg = <0x70000 0x10000>;
-		};
-		partition@80000 {
-			label = "qspi-linux";
-			reg = <0x80000 0x400000>;
-		};
-		partition@480000 {
+		partition@100000 {
 			label = "qspi-rootfs";
-			reg = <0x480000 0x1b80000>;
+			reg = <0x100000 0>;
 		};
 	};
 };
@@ -74,6 +67,14 @@
 &i2c1 {
 	status = "okay";
 	clock-frequency = <400000>;
+	/* GPIO expander */
+	gpioex: pca953x@41 {
+		compatible = "nxp,pca9536";
+		reg = <0x41>;
+		gpio-line-names = "USB_RESET", "VTT_SHDWN_N", "V_PRESENT", "DEBUG_PRESENT";
+		gpio-controller;
+		#gpio-cells = <2>;
+	};
 };
 
 &clkc {
diff --git a/configs/topic_miami_defconfig b/configs/topic_miami_defconfig
index 205e2e11ad..846d754da8 100644
--- a/configs/topic_miami_defconfig
+++ b/configs/topic_miami_defconfig
@@ -21,9 +21,11 @@ CONFIG_SYS_MEMTEST_END=0x18000000
 CONFIG_REMAKE_ELF=y
 CONFIG_SYS_CUSTOM_LDSCRIPT=y
 CONFIG_SYS_LDSCRIPT="arch/arm/mach-zynq/u-boot.lds"
+# CONFIG_BOOTSTD is not set
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_BOOTDELAY=0
 CONFIG_USE_PREBOOT=y
+# CONFIG_BOARD_LATE_INIT is not set
 CONFIG_CLOCKS=y
 CONFIG_SPL_MAX_SIZE=0x30000
 CONFIG_SPL_HAS_BSS_LINKER_SECTION=y
@@ -39,8 +41,12 @@ CONFIG_SYS_PROMPT="zynq-uboot> "
 CONFIG_SYS_MAXARGS=32
 CONFIG_SYS_PBSIZE=2077
 CONFIG_SYS_BOOTM_LEN=0x3c00000
+# CONFIG_CMD_BOOTEFI is not set
+# CONFIG_CMD_ELF is not set
+# CONFIG_CMD_SPL is not set
 CONFIG_CMD_THOR_DOWNLOAD=y
 CONFIG_THOR_RESET_OFF=y
+# CONFIG_CMD_SAVEENV is not set
 CONFIG_CMD_MEMTEST=y
 CONFIG_CMD_DFU=y
 CONFIG_CMD_I2C=y
@@ -48,6 +54,11 @@ CONFIG_CMD_MMC=y
 CONFIG_CMD_USB=y
 # CONFIG_CMD_SETEXPR is not set
 CONFIG_CMD_CACHE=y
+# CONFIG_CMD_EFICONFIG is not set
+CONFIG_CMD_SQUASHFS=y
+CONFIG_MTDIDS_DEFAULT="nor0=spi0.0"
+CONFIG_MTDPARTS_DEFAULT="mtdparts=spi0.0:0x100000(qspi-boot-bin),-(qspi-rootfs)"
+CONFIG_CMD_UBI=y
 CONFIG_OF_EMBED=y
 CONFIG_ENV_OVERWRITE=y
 CONFIG_SYS_RELOC_GD_ENV_ADDR=y
@@ -60,8 +71,12 @@ CONFIG_DM_I2C=y
 CONFIG_SYS_I2C_CADENCE=y
 CONFIG_MMC_SDHCI=y
 CONFIG_MMC_SDHCI_ZYNQ=y
+CONFIG_MTD=y
+CONFIG_DM_MTD=y
 CONFIG_SPI_FLASH_STMICRO=y
 # CONFIG_SPI_FLASH_USE_4K_SECTORS is not set
+CONFIG_SPI_FLASH_MTD=y
+CONFIG_MTD_UBI_BEB_LIMIT=0
 CONFIG_DEBUG_UART_ZYNQ=y
 CONFIG_ARM_DCC=y
 CONFIG_ZYNQ_SERIAL=y
@@ -77,3 +92,4 @@ CONFIG_USB_GADGET_PRODUCT_NUM=0x0300
 CONFIG_CI_UDC=y
 CONFIG_USB_GADGET_DOWNLOAD=y
 CONFIG_SYS_TIMER_COUNTS_DOWN=y
+CONFIG_LZ4=y
diff --git a/include/configs/topic_miami.h b/include/configs/topic_miami.h
index b0fcd240da..a620dd940d 100644
--- a/include/configs/topic_miami.h
+++ b/include/configs/topic_miami.h
@@ -29,8 +29,14 @@
 # define BOOT_TARGET_DEVICES_QSPI(func)
 #endif
 
+#ifdef CONFIG_CMD_UBIFS
+# define BOOT_TARGET_DEVICES_UBIFS(func) func(UBIFS, ubifs, 0, qspi-rootfs, qspi-rootfs)
+#else
+# define BOOT_TARGET_DEVICES_UBIFS(func)
+#endif
+
 #define BOOTENV_DEV_QSPI(devtypeu, devtypel, instance) \
-	"bootcmd_qspi=sf probe 0 0 0 && " \
+	"bootcmd_qspi=sf probe && " \
 		      "sf read ${scriptaddr} ${script_offset_f} ${script_size_f} && " \
 		      "echo QSPI: Trying to boot script at ${scriptaddr} && " \
 		      "source ${scriptaddr}; echo QSPI: SCRIPT FAILED: continuing...;\0"
@@ -40,6 +46,7 @@
 
 #define BOOT_TARGET_DEVICES(func) \
 	BOOT_TARGET_DEVICES_MMC(func) \
+	BOOT_TARGET_DEVICES_UBIFS(func) \
 	BOOT_TARGET_DEVICES_QSPI(func)
 
 #include <config_distro_bootcmd.h>
@@ -47,15 +54,15 @@
 #endif /* CONFIG_SPL_BUILD */
 
 /* Default environment */
-#ifndef CFG_EXTRA_ENV_SETTINGS
-#define CFG_EXTRA_ENV_SETTINGS	\
-	"script_size_f=0x40000\0"	\
-	"fdt_addr_r=0x1f00000\0"        \
-	"pxefile_addr_r=0x2000000\0"    \
-	"kernel_addr_r=0x2000000\0"     \
-	"ramdisk_addr_r=0x3100000\0"    \
+#define CFG_EXTRA_ENV_SETTINGS \
+	"scriptaddr=0x3000000\0"	\
+	"script_offset_f=0xf0000\0"	\
+	"script_size_f=0x10000\0"	\
+	"fdt_addr_r=0x1f00000\0"	\
+	"pxefile_addr_r=0x2000000\0"	\
+	"kernel_addr_r=0x2000000\0"	\
+	"ramdisk_addr_r=0x3100000\0"	\
 	BOOTENV
-#endif
 
 #include "zynq-common.h"
 
-- 
2.43.0

